url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/ACS.java

```diff
@@ -0,0 +1,1521 @@
+				for(AbstractAction a : shortlist)
+				{
+					AbstractOutputChunk o = i.getOutput(a.getID());
+					if(o != null)
+					{
+						sl.get(i).put(o, new GenericStochasticObject (o.getActivation()));
+						if(o.getID().equals(ChosenAction.getID()) && (max == null || o.getActivation() > max.getFinalSelectionMeasure()))
+							max = sl.get(i).get(o);
+					}
+				}
+				if(i instanceof InterfaceHandlesFeedback && i.Output.containsKey(ChosenAction.getID()))
+				{
+					((InterfaceHandlesFeedback)i).setFeedback(feedback);
+					if(i instanceof InterfaceHandlesNewInput)
+						((InterfaceHandlesNewInput)i).setNewInput(NewInput.values());
+					if(PERFORM_EXTRACTION && EXTRACTOR.checkExtractionCriterion(i))
+					{
+						AbstractRule R = RuleExtractor.extractRule(i,ChosenAction);
+						if(!RERRuleStore.contains(R) && !RERRuleStore.Children.contains(R))
+						{
+							((RefineableRule)R).setMatchAll(getMatchAllRule(R.getAction()));
+							RERRuleStore.add(R);
+						}
+						AbstractRule up = RERRuleStore.get(R.getCondition(), R.getAction());
+						if(up == null)
+							up = RERRuleStore.Children.get(R.getCondition(), R.getAction());
+						up.addTimeStamp(CurrentTimeStamp);
+					}
+					if(i instanceof InterfaceTrainable && PERFORM_BL_LEARNING)
+						((InterfaceTrainable)i).backwardPass();
+					
+					if(i instanceof InterfaceTracksMatchStatistics)
+					{
+						if(i instanceof InterfaceHasMatchCalculator)
+							((InterfaceTracksMatchStatistics)i).updateMatchStatistics(((InterfaceHasMatchCalculator)i).getMatchCalculator());
+						else
+							((InterfaceTracksMatchStatistics)i).updateMatchStatistics(MatchCalculator);
+					}
+				}
+			}
+		}
+		
+		BLImplicitModuleStore.setFeedback(feedback);
+		if(LEVEL_SELECTION_METHOD == ACS.LevelSelectionMethods.STOCHASTIC && ModuleUsed == ModuleTypes.BL)
+		{
+			BLImplicitModuleStore.updateMatchStatistics(MatchCalculator);
+		}
+		else
+		{
+			if(LEVEL_SELECTION_METHOD == ACS.LevelSelectionMethods.COMBINED)
+			{
+				LinkedList <InterfaceStochasticallySelectable> slc = new LinkedList <InterfaceStochasticallySelectable> ();
+				for(HashMap <AbstractOutputChunk, InterfaceStochasticallySelectable> h : sl.values())
+					slc.addAll(h.values());
+				if(MatchCalculator.isPositive(feedback, BLImplicitModuleStore.POSITIVE_MATCH_THRESHOLD))
+				{
+					BLImplicitModuleStore.setPM(BLImplicitModuleStore.getPM() + SELECTOR.getBoltzmannProbability(slc, max));
+				}
+				else
+					BLImplicitModuleStore.setNM(BLImplicitModuleStore.getNM() + SELECTOR.getBoltzmannProbability(slc, max));
+			}
+		}
+	}
+	
+	/**
+	 * Updates the match statistics for the rules in the rule store (if its type was used) and tries to refine those 
+	 * rules in the rule store whose type (i.e. RER/IRL/FR) was used.
+	 * @param feedback The feedback.
+	 * @param TimeStamp The current time stamp.
+	 */
+	private void updateRules (double feedback, long TimeStamp)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 82. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 743-1007 into a separate method.